src/main/org/firebirdsql/gds/ng/wire/SrpClient.java

```diff
@@ -0,0 +1,222 @@
+ *    1. Redistributions of source code must retain the above copyright notice,
+ *       this list of conditions and the following disclaimer.
+ *    2. Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *    3. The name of the author may not be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR IMPLIED
+ * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
+ * EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+ * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+ * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+ * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+// https://github.com/nakagami/pyfirebirdsql/blob/master/firebirdsql/srp.py
+
+package org.firebirdsql.gds.ng.wire;
+import java.util.Arrays;
+import java.math.BigInteger;
+import java.security.SecureRandom;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import javax.xml.bind.DatatypeConverter;
+
+/**
+ * @author <a href="mailto:nakagami@gmail.com">Hajime Nakagami</a>
+ */
+
+
+
+public final class SrpClient {
+    private static final int SRP_KEY_SIZE = 128;
+    private static final int SRP_SALT_SIZE = 32;
+    private static final BigInteger N = new BigInteger("E67D2E994B2F900C3F41F08F5BB2627ED0D49EE1FE767A52EFCD565CD6E768812C3E1E9CE8F0A8BEA6CB13CD29DDEBF7A96D4A93B55D488DF099A15C89DCB0640738EB2CBDD9A8F7BAB561AB1B0DC1C6CDABF303264A08D1BCA932D1F1EE428B619D970F342ABA9A65793B8B2F041AE5364350C16F735F56ECBCA87BD57B29E7", 16);
+    private static final BigInteger g = new BigInteger("2");
+
+    private static final BigInteger k = new BigInteger("1277432915985975349439481660349303019122249719989");
+
+    private BigInteger publicKey;   /* A */
+    private BigInteger privateKey;  /* a */
+    private byte[] sessionKey;      /* K */
+
+
+    private static class KeyPair {
+        private BigInteger pub, secret;
+        KeyPair(BigInteger pub, BigInteger secret) {
+            this.pub = pub;
+            this.secret = secret;
+        }
+        private BigInteger getPublicKey() {
+            return pub;
+        }
+        private BigInteger getPrivateKey() {
+            return secret;
+        }
+    }
+
+    private static BigInteger fromBigByteArray(byte[] b) {
+        return new BigInteger(DatatypeConverter.printHexBinary(b), 16);
```

Comment:
Sorry I didn't know the constructor, and I don't know which is better.
This conversion need keep positive, But I did'nt try BigInteger(int signum, byte[] magnitude) constructor, so I don't know the constructor keep positive or not.